feat(router): add navigationSource and restoredState to NavigationStart event

vsavkin opened the pull request:
Currently, NavigationStart there is no way to know if an navigation was triggered imperatively or via the location change. These two use cases should be handled differently for a variety of use cases (e.g., scroll position restoration). This PR adds a navigation source field and restored navigation id (passed to navigations triggered by a URL change).

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -208,6 +208,17 @@ export interface NavigationExtras {
 
 /** @stable */
 export declare class NavigationStart extends RouterEvent {
+    navigationSource?: NavigationSource;
```
NavigationSource is not exported but it's used as public api surface. is this intentional?

IgorMinar replied:
I recommend inlining the string literal union into the public api + renaming the property to navigationTrigger (source sounds like the source URL, which this is not).

IgorMinar commented on packages/router/src/events.ts:
```diff
@@ -42,6 +47,26 @@ export class RouterEvent {
  * @stable
  */
 export class NavigationStart extends RouterEvent {
+  /** @docsNotRequired */
```
why no docs?

IgorMinar commented on packages/router/src/events.ts:
```diff
@@ -42,6 +47,26 @@ export class RouterEvent {
+  /** @docsNotRequired */
+  navigationSource?: NavigationSource;
+
+  /** @docsNotRequired */
```
why no docs?

IgorMinar reviewed (changes requested):
this looks very good, we just need to tweak the public api a bit.

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -208,6 +208,17 @@ export interface NavigationExtras {
 export declare class NavigationStart extends RouterEvent {
+    navigationSource?: NavigationSource;
+    restoredState?: {
+        navigationId: number;
```
please add docs for the navigationId and what is its significance.

vsavkin commented:
* I added the missing docs. 
* I renamed NavigationSource into NavigationTrigger.
* I exported NavigationTrigger instead of inlining it. Mainly cause it is already used in a handful of places in the router itself. Inlining will result in a lot of duplication.

vsavkin committed:
feat(router): add navigationSource and restoredState to NavigationStart event

Currently, NavigationStart there is no way to know if an navigation was triggered imperatively or via the location change. These two use cases should be handled differently for a variety of use cases (e.g., scroll position restoration). This PR adds a navigation source field and restored navigation id (passed to navigations triggered by a URL change).

jasonaden commented:
I updated to remove the `NavigationTrigger` event from the external API. This is for documentation, so looking at this event on angular.io you will see the values you can use rather than the type name.

IgorMinar reviewed (commented):
this got submitted prematurely by accident. but all my comments were resolved so LGTM. thanks Victor.
